cryptoki/tests/slh_dsa.rs

```diff
@@ -0,0 +1,364 @@
+    session.verify_signature_init(&mechanism, public, &signature1)?;
+    session.verify_signature(&data)?;
+
+    // With Context + Deterministic Hedge
+    let context = [
+        0xEE, 0x0B, 0x3F, 0x67, 0x9F, 0xB5, 0x0F, 0x59, 0xAD, 0x31, 0x32, 0x8A, 0xAF, 0x4E, 0x70,
+        0x2C, 0xCF, 0x60, 0x92, 0xDA, 0x47, 0x94, 0xDC, 0xF0, 0x7C, 0x8, 0xEA, 0x27, 0x8B, 0x34,
+        0x22, 0x8A, 0x41,
+    ];
+    let mechanism = Mechanism::SlhDsa(SignAdditionalContext::new(
+        HedgeType::DeterministicRequired,
+        Some(&context),
+    ));
+
+    let signature2 = session.sign(&mechanism, private, &data)?;
+    let signature3 = session.sign(&mechanism, private, &data)?;
+    // Deterministic signature
+    assert_eq!(signature2, signature3);
+
+    session.verify(&mechanism, public, &data, &signature2)?;
+
+    // also using the new API
+    session.verify_signature_init(&mechanism, public, &signature2)?;
+    session.verify_signature(&data)?;
+
+    // the signature from previous step should fail to verify with different context
+    let result = session.verify(&mechanism, public, &data, &signature1);
+    assert!(result.is_err());
+    assert!(matches!(
+        result.unwrap_err(),
+        Error::Pkcs11(_, Function::Verify)
+    ));
+
+    // also using the new API
+    session.verify_signature_init(&mechanism, public, &signature1)?;
+    let result = session.verify_signature(&data);
+    assert!(result.is_err());
+    assert!(matches!(
+        result.unwrap_err(),
+        Error::Pkcs11(_, Function::VerifySignature)
+    ));
+
+    // Test convrting ParameterSet attributes back to algorithm specific values
+    let param_attribute = session
+        .get_attributes(public, &[AttributeType::ParameterSet])?
+        .remove(0);
+    let param: SlhDsaParameterSetType = if let Attribute::ParameterSet(num) = param_attribute {
+        num.into()
+    } else {
+        panic!("Expected ParameterSet attribute.");
+    };
+    assert_eq!(param, SlhDsaParameterSetType::SHA2_256F);
+
+    // delete keys
+    session.destroy_object(public)?;
+    session.destroy_object(private)?;
+
+    Ok(())
+}
+
+#[test]
+#[serial]
+fn slh_dsa_multipart() -> TestResult {
+    let (pkcs11, slot) = init_pins();
+    // PKCS#11 3.2 API is not supported by this token. Skip
+    if !pkcs11.is_fn_supported(Function::VerifySignature) {
+        /* return Ignore(); */
+        print!("SKIP: The PKCS#11 module does not support VerifySignature API");
+        return Ok(());
+    }
+
+    // open a session
+    let session = pkcs11.open_rw_session(slot)?;
+
+    // log in the session
+    session.login(UserType::User, Some(&AuthPin::new(USER_PIN.into())))?;
+
+    let mechanism = Mechanism::SlhDsaKeyPairGen;
+
+    // pub key template
+    let pub_key_template = vec![
+        Attribute::Token(true),
+        Attribute::ParameterSet(SlhDsaParameterSetType::SHA2_192S.into()),
+        Attribute::Verify(true),
+    ];
+
+    // priv key template
```

Comment:
To be honest I consider these all comments quite redundant since it's pretty obvious from the code on what's going on (for example, the comment is `priv key template` and on the next line the variable is called `priv_key_template`).

We don't pay you by the line count! :clown_face: :wink:

Reply:
Thanks for the review! Good point. I probably copied it from some previous test without much thinking. Will remove the obvious ones.